### PR TITLE
MLIR#842: fixed mixed layout in rock.conv2d()

### PR DESCRIPTION
**Pull request by sjw36**
* when 1x1 filters get tosa.reshape then converted to collapse(NCHW) -> expand(NHWC)
* added unit test

https://github.com/ROCmSoftwarePlatform/rocMLIR-internal/issues/842

### CODE REVIEW

File: mlir/lib/Conversion/TosaToRock/TosaToRock.cpp
Comment (krzysz00) [inline mlir/lib/Conversion/TosaToRock/TosaToRock.cpp:606]: Nit: we could get rid of `lres` and make this `if (!colOp || !checkCollapse(colOp)) { return failure(); }`

File: mlir/lib/Conversion/TosaToRock/TosaToRock.cpp
Comment (krzysz00) [inline mlir/lib/Conversion/TosaToRock/TosaToRock.cpp:611]: Do we want this pattern for input and output tensors too?

**Review by krzysz00 — COMMENTED** [review]
Overall seems reasonable, nothing I commented is a blocker.

File: mlir/lib/Conversion/TosaToRock/TosaToRock.cpp
Comment (sjw36) [reply mlir/lib/Conversion/TosaToRock/TosaToRock.cpp:606]: Still need it for the user loop, there could theoretically be multiple users.

File: mlir/lib/Conversion/TosaToRock/TosaToRock.cpp
Comment (sjw36) [reply mlir/lib/Conversion/TosaToRock/TosaToRock.cpp:611]: Limiting to filters since they are probably the only 1x1 tensors we will see.

File: mlir/lib/Conversion/TosaToRock/TosaToRock.cpp
Comment (krzysz00) [reply mlir/lib/Conversion/TosaToRock/TosaToRock.cpp:606]: Ah, missed that

We could still do the early return to save on a level of nesting

**Review by krzysz00 — APPROVED** [review]
Approved since there's one real nitpick left (inverting the collapse check if statement) but there's no reason to make you wait for me to hit approve after that's fixed.

**Pull request merged**
